File: engine/crates/blocus-python/python_tests/test_edge_cases.py
from __future__ import annotations

import json

import pytest

import blocus_engine as be


GAME_ID = "00000000-0000-0000-0000-000000000900"
PLAYER_1 = "00000000-0000-0000-0000-000000000001"
PLAYER_2 = "00000000-0000-0000-0000-000000000002"
PLAYER_3 = "00000000-0000-0000-0000-000000000003"
PLAYER_4 = "00000000-0000-0000-0000-000000000004"


def uuid(value: int) -> str:
    return f"00000000-0000-0000-0000-{value:012d}"


def official_turn_order() -> list[be.PlayerColor]:
    return [
        be.PlayerColor.BLUE,
        be.PlayerColor.YELLOW,
        be.PlayerColor.RED,
        be.PlayerColor.GREEN,
    ]


def three_player_state() -> tuple[be.BlocusEngine, be.GameState]:
    engine = be.BlocusEngine()
    shared = be.SharedColorTurn(
        be.PlayerColor.GREEN,
        [PLAYER_1, PLAYER_2, PLAYER_3],
    )
    slots = be.PlayerSlots.three_player(
        [
            (be.PlayerColor.BLUE, PLAYER_1),
            (be.PlayerColor.YELLOW, PLAYER_2),
            (be.PlayerColor.RED, PLAYER_3),
        ],
        shared,
    )
    config = be.GameConfig(
        GAME_ID,
        be.GameMode.THREE_PLAYER,
        be.ScoringMode.BASIC,
        official_turn_order(),
        slots,
    )

    return engine, engine.initialize_game(config)


def two_player_state() -> tuple[be.BlocusEngine, be.GameState]:
    engine = be.BlocusEngine()
    config = be.GameConfig.two_player(
        GAME_ID,
        PLAYER_1,
        PLAYER_2,
        be.ScoringMode.BASIC,
    )

    return engine, engine.initialize_game(config)


def state_with_turn(
    state: be.GameState,
    current_color: str,
    shared_color_turn_index: int = 0,
    passed_mask: int = 0,
) -> be.GameState:
    data = json.loads(state.to_json())
    data["turn"]["current_color"] = current_color
    data["turn"]["shared_color_turn_index"] = shared_color_turn_index
    data["turn"]["passed_mask"] = passed_mask

    return be.GameState.from_json(json.dumps(data))


def test_three_player_shared_color_scheduled_ownership_is_exposed_by_state_transition() -> None:
    engine, state = three_player_state()
    state = state_with_turn(state, "green", shared_color_turn_index=1)

    command = be.PlaceCommand(
        command_id=uuid(1),
        game_id=GAME_ID,
        player_id=PLAYER_2,
        color=be.PlayerColor.GREEN,
        piece_id=0,
        orientation_id=0,
        row=19,
        col=0,
    )

    result = engine.apply(state, command)

    assert result.next_state.current_color == be.PlayerColor.BLUE
    assert result.next_state.version == state.version + 1
    assert result.response.kind == be.DomainResponseKind.MOVE_APPLIED


def test_illegal_shared_color_move_by_wrong_shared_player_is_rejected() -> None:
    engine, state = three_player_state()
    state = state_with_turn(state, "green", shared_color_turn_index=1)

    command = be.PlaceCommand(
        command_id=uuid(2),
        game_id=GAME_ID,
        player_id=PLAYER_1,
        color=be.PlayerColor.GREEN,
        piece_id=0,
        orientation_id=0,
        row=19,
        col=0,
    )

    with pytest.raises(be.RuleViolationError) as captured:
        engine.apply(state, command)

    assert "PlayerDoesNotControlColor" in str(captured.value)


def test_legal_shared_color_move_by_correct_shared_player_is_accepted() -> None:
    engine, state = three_player_state()
    state = state_with_turn(state, "green", shared_color_turn_index=2)

    command = be.PlaceCommand(
        command_id=uuid(3),
        game_id=GAME_ID,
        player_id=PLAYER_3,
        color=be.PlayerColor.GREEN,
        piece_id=0,
        orientation_id=0,
        row=19,
        col=0,
    )

    result = engine.apply(state, command)

    assert result.next_state.version == state.version + 1
    assert result.next_state.board_is_empty is False
    assert [event.kind for event in result.events] == [
        be.DomainEventKind.MOVE_APPLIED,
        be.DomainEventKind.TURN_ADVANCED,
    ]


def test_pass_hash_is_recomputed_after_successful_pass() -> None:
    engine, state = two_player_state()

    data = json.loads(state.to_json())
    data["board"]["yellow"]["lanes"] = ["4294967299", "0", "0", "0", "0"]
    data["board"]["yellow"]["count"] = 4

    blocked_blue = be.GameState.from_json(json.dumps(data))

    command = be.PassCommand(
        command_id=uuid(4),
        game_id=GAME_ID,
        player_id=PLAYER_1,
        color=be.PlayerColor.BLUE,
    )

    result = engine.apply(blocked_blue, command)

    assert result.next_state.version == blocked_blue.version + 1
    assert result.next_state.hash != 0
    assert result.next_state.hash == be.GameState.from_json(result.next_state.to_json()).hash


def test_four_player_rejects_illegal_non_clockwise_permutation() -> None:
    with pytest.raises(be.InputError) as captured:
        be.GameConfig.four_player(
            GAME_ID,
            [
                (be.PlayerColor.BLUE, PLAYER_1),
                (be.PlayerColor.YELLOW, PLAYER_2),
                (be.PlayerColor.RED, PLAYER_3),
                (be.PlayerColor.GREEN, PLAYER_4),
            ],
            be.ScoringMode.BASIC,
            [
                be.PlayerColor.BLUE,
                be.PlayerColor.RED,
                be.PlayerColor.YELLOW,
                be.PlayerColor.GREEN,
            ],
        )

    assert "InvalidGameConfig" in str(captured.value)


def test_json_board_masks_with_padding_bits_are_rejected() -> None:
    _, state = two_player_state()
    data = json.loads(state.to_json())

    data["board"]["blue"]["lanes"] = [str(1 << 20), "0", "0", "0", "0"]
    data["board"]["blue"]["count"] = 1

    with pytest.raises(be.EngineError) as captured:
        be.GameState.from_json(json.dumps(data))

    assert "CorruptedState" in str(captured.value)


def test_json_board_masks_with_overlapping_colors_are_rejected() -> None:
    _, state = two_player_state()
    data = json.loads(state.to_json())

    data["board"]["blue"]["lanes"] = ["1", "0", "0", "0", "0"]
    data["board"]["blue"]["count"] = 1
    data["board"]["yellow"]["lanes"] = ["1", "0", "0", "0", "0"]
    data["board"]["yellow"]["count"] = 1

    with pytest.raises(be.EngineError) as captured:
        be.GameState.from_json(json.dumps(data))

    assert "CorruptedState" in str(captured.value)

File: engine/crates/blocus-core/tests/test_edge_cases.rs
use blocus_core::{
    BlocusEngine, BoardIndex, BoardMask, Command, CommandId, DomainError, GameConfig, GameId,
    GameMode, OrientationId, PassCommand, PieceId, PlaceCommand, PlayerColor, PlayerId,
    PlayerSlots, RuleViolation, ScoringMode, SharedColorTurn, TurnOrder, ZobristHash,
};
use uuid::Uuid;

fn uuid(value: u128) -> Uuid {
    Uuid::from_u128(value)
}

fn game_id(value: u128) -> GameId {
    GameId::from_uuid(uuid(value))
}

fn player_id(value: u128) -> PlayerId {
    PlayerId::from_uuid(uuid(value))
}

fn command_id(value: u128) -> CommandId {
    CommandId::from_uuid(uuid(value))
}

fn piece_id(value: u8) -> PieceId {
    PieceId::try_new(value).unwrap_or_else(|_| panic!("piece id {value} should be valid"))
}

fn orientation_id(value: u8) -> OrientationId {
    OrientationId::try_new(value)
        .unwrap_or_else(|_| panic!("orientation id {value} should be valid"))
}

fn index(row: u8, col: u8) -> BoardIndex {
    BoardIndex::from_row_col(row, col)
        .unwrap_or_else(|_| panic!("row {row}, col {col} should be valid"))
}

fn three_player_config_shared_green() -> GameConfig {
    let shared = SharedColorTurn::try_new(
        PlayerColor::Green,
        [player_id(1), player_id(2), player_id(3)],
    )
    .unwrap_or_else(|_| panic!("shared turn should be valid"));

    let slots = PlayerSlots::three_player(
        [
            (PlayerColor::Blue, player_id(1)),
            (PlayerColor::Yellow, player_id(2)),
            (PlayerColor::Red, player_id(3)),
        ],
        shared,
    )
    .unwrap_or_else(|_| panic!("three-player slots should be valid"));

    GameConfig::try_new(
        game_id(100),
        GameMode::ThreePlayer,
        ScoringMode::Basic,
        TurnOrder::OFFICIAL_FIXED,
        slots,
    )
    .unwrap_or_else(|_| panic!("three-player config should be valid"))
}

fn two_player_config() -> GameConfig {
    let slots = PlayerSlots::two_player(player_id(1), player_id(2))
        .unwrap_or_else(|_| panic!("two-player slots should be valid"));

    GameConfig::try_new(
        game_id(100),
        GameMode::TwoPlayer,
        ScoringMode::Basic,
        TurnOrder::OFFICIAL_FIXED,
        slots,
    )
    .unwrap_or_else(|_| panic!("two-player config should be valid"))
}

fn four_player_slots() -> PlayerSlots {
    PlayerSlots::four_player([
        (PlayerColor::Blue, player_id(1)),
        (PlayerColor::Yellow, player_id(2)),
        (PlayerColor::Red, player_id(3)),
        (PlayerColor::Green, player_id(4)),
    ])
    .unwrap_or_else(|_| panic!("four-player slots should be valid"))
}

#[test]
fn edge_case_three_player_shared_color_accepts_wrong_scheduled_player() {
    let engine = BlocusEngine::new();
    let mut state = engine.initialize_game(three_player_config_shared_green());

    state.turn = blocus_core::TurnState::from_parts(PlayerColor::Green, 0, 0);

    let command = Command::Place(PlaceCommand {
        command_id: command_id(1),
        game_id: state.game_id,
        player_id: player_id(2),
        color: PlayerColor::Green,
        piece_id: piece_id(0),
        orientation_id: orientation_id(0),
        anchor: index(19, 0),
    });

    assert_eq!(
        engine.apply(&state, command),
        Err(DomainError::from(RuleViolation::PlayerDoesNotControlColor))
    );
}

#[test]
fn edge_case_four_player_accepts_non_clockwise_turn_order() {
    let non_clockwise_order = TurnOrder::try_new([
        PlayerColor::Blue,
        PlayerColor::Red,
        PlayerColor::Yellow,
        PlayerColor::Green,
    ])
    .unwrap_or_else(|_| panic!("permutation is structurally valid"));

    assert_eq!(
        GameConfig::try_new(
            game_id(101),
            GameMode::FourPlayer,
            ScoringMode::Basic,
            non_clockwise_order,
            four_player_slots(),
        ),
        Err(blocus_core::InputError::InvalidGameConfig)
    );
}

#[test]
fn edge_case_successful_pass_hash_is_not_recomputed() {
    let engine = BlocusEngine::new();
    let mut state = engine.initialize_game(two_player_config());

    state.board.place_mask(
        PlayerColor::Yellow,
        BoardMask::from_index(index(0, 0))
            .union(BoardMask::from_index(index(0, 1)))
            .union(BoardMask::from_index(index(1, 0)))
            .union(BoardMask::from_index(index(1, 1))),
    );

    assert_eq!(
        engine.has_any_valid_move(&state, player_id(1), PlayerColor::Blue),
        Ok(false)
    );

    let result = engine
        .apply(
            &state,
            Command::Pass(PassCommand {
                command_id: command_id(1),
                game_id: state.game_id,
                player_id: player_id(1),
                color: PlayerColor::Blue,
            }),
        )
        .unwrap_or_else(|error| panic!("pass should be legal: {error}"));

    assert_eq!(
        result.next_state.hash,
        blocus_core::compute_hash_full(&result.next_state)
    );
    assert_ne!(result.next_state.hash, ZobristHash::ZERO);
}

#[test]
fn edge_case_corrupted_padding_bit_board_state_is_accepted() {
    let engine = BlocusEngine::new();
    let mut state = engine.initialize_game(two_player_config());

    let padding_bit_mask = BoardMask::from_lanes([1u128 << 20, 0, 0, 0, 0]);
    state.board.place_mask(PlayerColor::Blue, padding_bit_mask);
    state.hash = blocus_core::compute_hash_full(&state);

    assert!(!state.board.occupied(PlayerColor::Blue).is_playable_subset());

    assert_eq!(
        engine.has_any_valid_move(&state, player_id(1), PlayerColor::Blue),
        Err(DomainError::from(blocus_core::EngineError::CorruptedState))
    );
}
